Make aircraft occupy cells when landed.

tovl opened the pull request:
This makes aircraft occupy space on the `ActorMap` when landed. This prevents other ground units from moving into the same space.

Fixes #4605.
Fixes #2018.
Fixes #15131
Fixes #14114
Fixes #14867

This makes fixed-wing aircraft check for blocking actors as well. If they find the repair depot (or any other landing spot) blocked, they will circle and wait for it to be unblocked. Aircraft will try to reserve the required cells when landing is initiated.

tovl committed:
Make aircraft occupy cells when landed.

pchote commented:
Can you comment on how this interacts with crates and mines?

teinarss commented on mods/cnc/rules/aircraft.yaml:
```diff
@@ -68,6 +68,7 @@ HELI:
 	Aircraft:
 		TurnSpeed: 7
 		Speed: 180
+		LandableTerrainTypes: Clear,Rough,Road,Ore,Beach,Tiberium,BlueTiberium
```
Should prolly be added to the transport heli too, Chinook.

tovl replied:
This was copy-pasted from the transport heli; See Line 18.

tovl commented:
> Can you comment on how this interacts with crates and mines?

Helicopters cannot land on mines or crates. This is true in bleed as well.

pchote commented:
Is there any prospect for fixing that here? Aircraft not interacting with the ActorMap was always the main blocker for supporting those.

tovl commented:
I think that should be handled in a follow-up PR.

reaperrr commented on mods/cnc/rules/aircraft.yaml:
```diff
@@ -68,6 +68,7 @@ HELI:
 	Aircraft:
 		TurnSpeed: 7
 		Speed: 180
+		LandableTerrainTypes: Clear,Rough,Road,Ore,Beach,Tiberium,BlueTiberium
```
All aircraft in TD except transport helis have `LandWhenIdle: false` and can only land on helipads, and the terrain type shouldn't matter there, so these yaml changes should be unnecessary.

tovl replied:
Unfortunately, that is not true. In order to fix #2018 aircraft will have to explicitly check whether they can land on a given cell even when called from ReturnToBase. when calling CanLand they will then also check the underlying terrain. I admit this is a bit hacky. There is probably a better solution to this.

Even so, I think helicopters should at least in theory be able to land outside of the helipad even if currently there is no way to order them to do so. They could in the original and conceptually it makes sense.

tovl replied:
Ok. I've added an extra check in `CanLand` so this hack is not necessary anymore. Removed `LandableTerrainTypes` from non-vtol craft. I've left it in for helicopters. I guess that's just a matter of taste. I will remove it if you find it really objectionable.

reaperrr replied:
Well, 'objectionable' sounds a bit too harsh, we just try to avoid keeping lines around that won't have any effect on that actor.
That being said, as a compromise you could move this to the `^Helicopter` default instead.

reaperrr commented on mods/ra/rules/aircraft.yaml:
```diff
@@ -118,6 +118,7 @@ MIG:
 		Speed: 223
 		RepulsionSpeed: 40
 		MaximumPitch: 56
+		LandableTerrainTypes: Clear,Rough,Road,Beach
```
All aircraft in RA except transport helis have `LandWhenIdle: false` and can only land on airfields/helipads or repair depots, and the terrain type shouldn't matter there, so these yaml changes should be unnecessary.

tovl replied:
Ok. I've added an extra check in `CanLand` so this hack is not necessary anymore. Removed `LandableTerrainTypes` from non-vtol craft. I've left it in for helicopters. I guess that's just a matter of taste. I will remove it if you find it really objectionable.

reaperrr replied:
See my comment on cnc above.

tovl committed:
Add crushing logic to aircraft.

tovl commented:
So, I've decided to add the crushing logic to aircraft anyway. This will allow transport helicopters to interact with crates and mines. It will also allow them to squash infantry unlucky enough to stand right under them when landing.

matjaeck commented:
Is this intended to fix helicopters, too? For example 2 helis on one pad or vehicle and heli on fix.

tovl commented:
It should work for all types of aircraft.

tovl commented:
Helicopters should be fixed now.

matjaeck commented:
Great, I'll also test it against these issues, but not tonight: #14114, #14867, #15131

tovl commented:
Update:
- Fixed chinook being unable to unload cargo due to occupying space.
- Added missing `CrushDamageTypes` yaml parameter.
- aircraft can now nudge friendly units out of the way when landing.

tovl commented:
Rebased.

tovl commented:
Rebased.

matjaeck commented:
:+1:

reaperrr commented:
Transport heli constantly lands and takes off again.

matjaeck commented:
> Transport heli constantly lands and takes off again.

I assumed this is the already fixed #16368.

reaperrr commented:
I rebased onto latest bleed, but I still get that bug, so this PR appears to regress it again.

tovl commented:
Fixed.

reaperrr commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -640,6 +681,35 @@ public void SetPosition(Actor self, WPos pos)
 				OnCruisingAltitudeReached();
 			else if (!isCruising && cruising)
 				OnCruisingAltitudeLeft();
+			FinishedMoving(self);
```
I'd prefer an empty line above this for better readability.

tovl replied:
Fixed.

reaperrr reviewed (dismissed):
Looks good to me apart from minor nits

reaperrr commented:
Needs a rebase, too.

tovl commented:
Rebased.

reaperrr commented:
@tovl I guess you overlooked it (understandable since it's quite a bit up in the comments):

Regarding the `LandableTerrain` of RA/TD helicopters (and nearly all TS aircraft):
> Well, 'objectionable' sounds a bit too harsh, we just try to avoid keeping lines around that won't have any effect on that actor.
That being said, as a compromise you could move this to the `^Helicopter` default instead.

tovl commented:
Yes, I missed that. I've removed these lines for now, because as you said they do not have any effect currently.

reaperrr commented:
```
[18:44] <+reaperrr> pchote: what was the technical reason again for preferably not using writeable public booleans, ints and so on?
[18:49] <+pchote> in what context?
[18:52] <+reaperrr> https://github.com/OpenRA/OpenRA/pull/16315/commits/30e77af6049e92d34adf22f7c4ed7a5e2ac47c57#diff-e8b231bb90082b00e18a95d9ae6a7ef4R187
[18:54] <+pchote> basically want to expose as small a public surface area as possible
[18:56] <+pchote> that LandingInitiated doesn't seem to be used at all?
```

pchote commented:
`LandingCell` could perhaps be kept private by passing the cell as an argument to `AddInfluence`.

tovl commented:
Fixed.

reaperrr commented:
In addition to #16417, the changes here - in particular to Land/HeliLand - are also a dependency for updating #16365, as well as a few of my local PRs, so it would be good to get this merged soon (as long as there are no remaining issues, of course).

reaperrr commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -528,6 +548,41 @@ public bool CanRepairAt(Actor host)
+		bool IsBlockedBy(Actor self, Actor otherActor, Actor ignoreActor)
+		{
+			// We are not blocked by the actor we are ignoring.
+			if (otherActor == ignoreActor)
```
If you change this to `if (otherActor == self || otherActor == ignoreActor)` you can drop the work-arounds...

tovl replied:
Yep, I already did this is #16417. But it makes more sense to include it here. Changed.

reaperrr commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -575,7 +630,7 @@ protected virtual void OnBecomingIdle(Actor self)
 			else if (!Info.CanHover && !atLandAltitude)
 				self.QueueActivity(new FlyCircle(self, -1, Info.IdleTurnSpeed > -1 ? Info.IdleTurnSpeed : TurnSpeed));
-			else if (atLandAltitude && !CanLand(self.Location) && ReservedActor == null)
+			else if (atLandAltitude && !CanLand(self.Location, self) && ReservedActor == null)
```
...here...

tovl replied:
Fixed.

reaperrr commented on OpenRA.Mods.Common/Traits/Cargo.cs:
```diff
@@ -213,7 +213,7 @@ public bool CanUnload(bool immediate = false)
 			}
 
-			return !IsEmpty(self) && (aircraft == null || aircraft.CanLand(self.Location))
+			return !IsEmpty(self) && (aircraft == null || aircraft.CanLand(self.Location, self))
```
...and here.

tovl replied:
Fixed.

reaperrr reviewed (changes requested):
Thinking about it, the aircraft should *never* block itself from landing.

pchote commented on OpenRA.Mods.Common/Activities/Air/Fly.cs:
```diff
@@ -105,6 +105,9 @@ public override Activity Tick(Actor self)
 				soundPlayed = true;
 			}
 
+			if (self.IsAtGroundLevel())
```
Can you please add a comment here like `// We are taking off, so remove ground influence` to help reduce confusion?

tovl replied:
Done.

pchote commented on OpenRA.Mods.Common/Activities/Air/HeliFly.cs:
```diff
@@ -98,6 +98,9 @@ public override Activity Tick(Actor self)
 				soundPlayed = true;
 			}
 
+			if (self.IsAtGroundLevel())
```
Likewise here.

tovl replied:
Done.

pchote commented on OpenRA.Mods.Common/Activities/Air/HeliLand.cs:
```diff
@@ -19,26 +19,42 @@ public class HeliLand : Activity
+				var landingCell = self.Location;
+				if (!aircraft.CanLand(landingCell, ignoreActor))
+				{
+					self.NotifyBlocker(landingCell);
```
Pinging the blocker every tick isn't great for performance. Please add a delay between retries like you do in `Land`.

tovl replied:
Done.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -493,7 +498,13 @@ public int MovementSpeed
-		public Pair<CPos, SubCell>[] OccupiedCells() { return NoCells; }
+		public Pair<CPos, SubCell>[] OccupiedCells()
+		{
+			if (!self.IsAtGroundLevel())
```
`OccupyCells` can be called by arbitrary things at arbitrary times, so we should still be returning `NoCells` during normal flight to avoid bogosity.

tovl replied:
Okay. Tied this to `RemoveInfluence` to mirror `AddInfluence`.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -506,13 +517,22 @@ public WVec FlyStep(int speed, int facing)
-			if (self.World.ActorMap.AnyActorsAt(cell))
-				return false;
+			foreach (var otherActor in self.World.ActorMap.GetActorsAt(cell))
+			{
```
Style nit: these braces can be omitted

tovl replied:
Fixed.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -506,13 +517,22 @@ public WVec FlyStep(int speed, int facing)
+			}
+
+			foreach (var otherActor in self.World.ActorMap.GetActorsAt(cell))
+			{
```
These ones too.

tovl replied:
Fixed.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -528,6 +548,41 @@ public bool CanRepairAt(Actor host)
+			// If self is null, we don't have a real actor - we're just checking what would happen theoretically.
+			// In such a scenario - we'll just assume any other actor in the cell will block us by default.
+			// If we have a real actor, we can then perform the extra checks that allow us to avoid being blocked.
+			if (self == null)
```
This is inside `Aircraft` (not `AircraftInfo`), so `self` is guaranteed by convention to be non-null.

tovl replied:
Oops. That's what I get for shamelessly copy-pasting. Fixed.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -615,16 +670,61 @@ public void SetPosition(Actor self, WPos pos)
+
+			var notifiers = actors.SelectMany(a => a.TraitsImplementing<INotifyCrushed>().Select(t => new TraitPair<INotifyCrushed>(a, t)));
+			foreach (var notifyCrushed in notifiers)
+				notifyCrushed.Trait.OnCrush(notifyCrushed.Actor, self, Info.Crushes);
```
We need to call `WarnCrush` first, probably when starting to land to match `Mobile.EnteringCell`, so that enemy units have a chance to run away.

tovl replied:
Added `EnteringCell` method to `Aircraft` and called upon landing initiation.

matjaeck replied:
Something is not working, they always die.

tovl replied:
Should be fixed now.

pchote reviewed (commented):
Looking good, but a few comments:

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9e39b86618fd1fa5c51285350436785cbe0d2ce1...324d6479288abbd7d7b659abd2c6f4a65dca347c
